include/rocksdb/env_encrypt2.h

```diff
@@ -0,0 +1,473 @@
+//  copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+//
+// env_encryption.cc copied to this file then modified.
+
+#ifdef ROCKSDB_OPENSSL_AES_CTR
+#ifndef ROCKSDB_LITE
+
+#include <algorithm>
+#include <cctype>
+#include <iostream>
+
+#include "openssl/aes.h"
+#include "openssl/evp.h"
+#include "openssl/rand.h"
+#include "rocksdb/env_encryption.h"
+#include "util/aligned_buffer.h"
+#include "util/coding.h"
+#include "util/library_loader.h"
+#include "util/random.h"
+
+#endif
+
+namespace ROCKSDB_NAMESPACE {
+
+#ifndef ROCKSDB_LITE
+
+struct Sha1Description {
```

Comment:
I think this adds way more to the public API than is necessary.  I think most of this code can be moved into the CC file and not in the public API.

If you did the move, could you then methods to add/set new keys to the base EncryptedEnv that return NotSupported?  

As far as the public portion of the API, it seems like the only change is adding the changing of the keys.  Is there a reason to expose this as a Sha1Description versus something that is managed internally?  Or have a more abstract structure that you can use (like potentially a key name, nonce, IV, ...)?

More comments to come later...

Reply:
@mrambacher Not saying I am correct ... I was trying to make more things available to future developers after having to fight, then restructure, the original env_encryption.cc/.h.  Maybe the compromise is to move these structures into an implementation .h within the env directory so they can be reused / extended easily ... while still not cluttering the public API files?  It feels wrong to me to bury them in the .cc.  That makes them useless to the next coder.

Reply:
@mrambacher I have moved template implementations to .cc file.  Added env_encrypt2_impl.h for one class that is not needed for public API, but might be reusable in other implementations.